### PR TITLE
apply_generic_arguments: replace runtime checks with assertions

### PR DESCRIPTION
**Pull request by elazarg**
Part of #2272.

The only place where the number of type arguments in type application might not match is when the user does that; this is explicitly checked in `visit_type_application`. Other places involve type inference, which must be correct "by construction" and if it isn't, that's a bug we want to catch, instead of returning AnyType.

It also allows us to remove casts.

cc @ddfisher

### CODE REVIEW

File: mypy/subtypes.py [hunk 1]
```diff
@@ -333,7 +333,7 @@ def unify_generic_callable(type: CallableType, target: CallableType,
         return None
     msg = messages.temp_message_builder()
     applied = mypy.applytype.apply_generic_arguments(type, inferred_vars, msg, context=target)
-    if msg.is_errors() or not isinstance(applied, CallableType):
+    if msg.is_errors():
```
Comment (elazarg) [inline mypy/subtypes.py]: Perhaps this should be an assertion too?

Comment (gvanrossum) [inline mypy/subtypes.py]: I doubt it, unless you can prove that apply_generic_arguments() cannot produce any errors.

File: mypy/applytype.py [hunk 2]
```diff
@@ -3,25 +3,23 @@
...
+                            msg: MessageBuilder, context: Context) -> CallableType:
     """Apply generic type arguments to a callable type.
 
     For example, applying [int] to 'def [T] (T) -> T' results in
     'def (int) -> int'.
 
     Note that each type can be None; in this case, it will not be applied.
+    Can only be None when len(tvars) != len(types)
     """
+    assert len(callable.variables) == len(types)
```
Comment (gvanrossum) [inline mypy/applytype.py]: I'd move the assert after the assignment.

Comment (gvanrossum) [inline mypy/applytype.py]: I'd move the assert after the assignment.

File: mypy/subtypes.py (hunk 1 above)
Comment (elazarg) [inline mypy/subtypes.py]: Should it?

Comment (rwbarton) [inline mypy/subtypes.py]: It definitely can produce errors, since it checks type variable bounds/allowed values.

**Comment by gvanrossum** [comment]
Thanks!

**Pull request merged**